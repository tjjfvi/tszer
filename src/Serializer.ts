import { Enga, enga } from "enga";
import { AsyncEnga, asyncEnga } from "enga/async";

export interface SerializeResult {
  length: number,
  write: (buffer: Buffer, offset: number) => Enga<void>
};

export interface DeserializeResult<T> {
  length: number,
  value: T,
}

export type SerializeFunc<T> = (value: T) => Enga<SerializeResult>
export type DeserializeFunc<T> = (buffer: Buffer, offset: number) => AsyncEnga<DeserializeResult<T>>

export interface SerializerArgs<T> {
  serialize: SerializeFunc<T>,
  deserialize: DeserializeFunc<T>,
}

export type ConstLengthSerializeFunc<T> = (value: T, buffer: Buffer, offset: number) => void
export type ConstLengthDeserializeFunc<T> = (buffer: Buffer, offset: number) => T | Promise<T>

export interface ConstLengthSerializerArgs<T> {
  length: number,
  serialize: ConstLengthSerializeFunc<T>,
  deserialize: ConstLengthDeserializeFunc<T>,
}

export class Serializer<T> {

  serialize: SerializeFunc<T>;
  deserialize: DeserializeFunc<T>;

  constructor(args: ConstLengthSerializerArgs<T>)
  constructor(args: SerializerArgs<T>)
  constructor(args: SerializerArgs<T> | ConstLengthSerializerArgs<T>) {
    if ("length" in args) {
      const { length } = args;
      this.serialize = value => enga(() => ({
        length,
        write: (buffer, offset) => enga(() => args.serialize(value, buffer, offset)),
      }))
      this.deserialize = (buffer, offset) => asyncEnga(async () => ({
        length,
        value: await args.deserialize(buffer, offset),
      }));
    } else {
      this.serialize = args.serialize;
      this.deserialize = args.deserialize;
    }
  }

  map<U>({
    serialize,
    deserialize,
  }: {
    serialize: (value: U) => T,
    deserialize: (value: T) => U | Promise<U>,
  }) {
    return new Serializer<U>({
      serialize: value => enga(
        () => serialize(value),
        this.serialize,
      ),
      deserialize: (buffer: Buffer, offset: number) => asyncEnga(
        this.deserialize(buffer, offset),
        async result => ({ ...result, value: await deserialize(result.value) })
      ),
    })
  }

  errorOnNull(message: string) {
    return this.map<NonNullable<T>>({
      serialize: v => v,
      deserialize: v => {
        if (v == null)
          throw new Error(message);
        return v as NonNullable<T>;
      }
    })
  }

  static serialize<T>(serializer: Serializer<T>, value: T) {
    const { length, write } = serializer.serialize(value).execute();
    const buffer = Buffer.alloc(length);
    write(buffer, 0).execute();
    return buffer;
  }

  static async deserialize<T>(serializer: Serializer<T>, buffer: Buffer) {
    return (await serializer.deserialize(buffer, 0).execute()).value;
  }

}