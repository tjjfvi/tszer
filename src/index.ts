
// @create-index {"mode":"*"}

export * from './Serializer';
export * from './array';
export * from './buffer';
export * from './concat';
export * from './constant';
export * from './double';
export * from './flags';
export * from './float';
export * from './int';
export * from './optionalBank';
export * from './string';
export * from './tree';
export * from './uint';

