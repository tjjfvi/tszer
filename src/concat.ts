import { Serializer, DeserializeResult } from "./Serializer";

export type ConcatArg<Prev, Cur> = Serializer<Cur> | ((prev: Prev) => Serializer<Cur>);

function _concat(...args: ConcatArg<any[], any>[]): Serializer<any[]> {
  return new Serializer({
    serialize: value => value.map((v, i) => {
      let arg = args[i];
      if (typeof arg === "function")
        arg = arg(value.slice(0, i));
      return arg.serialize(v);
    }).reduce((acc, cur) => ({
      length: acc.length + cur.length,
      write: (buf, off) => {
        acc.write(buf, off);
        cur.write(buf, off + acc.length);
      }
    }), { length: 0, write: () => { } }),
    deserialize: (buf, off) => args.reduce<DeserializeResult<any[]>>((acc, arg) => {
      if (typeof arg === "function")
        arg = arg(acc.value);
      let result = arg.deserialize(buf, off + acc.length);
      return {
        length: acc.length + result.length,
        value: [...acc.value, result.value],
      }
    }, { length: 0, value: [] }),
  })
}

export function concat<A>(a: ConcatArg<[], A>): Serializer<[A]>
export function concat<A, B>(a: ConcatArg<[], A>, b: ConcatArg<[A], B>): Serializer<[A, B]>
export function concat<A, B, C>(a: ConcatArg<[], A>, b: ConcatArg<[A], B>, c: ConcatArg<[A, B], C>): Serializer<[A, B, C]>
export function concat<A, B, C, D>(a: ConcatArg<[], A>, b: ConcatArg<[A], B>, c: ConcatArg<[A, B], C>, d: ConcatArg<[A, B, C], D>): Serializer<[A, B, C, D]>
export function concat<A, B, C, D, E>(
  a: ConcatArg<[], A>,
  b: ConcatArg<[A], B>,
  c: ConcatArg<[A, B], C>,
  d: ConcatArg<[A, B, C], D>,
  e: ConcatArg<[A, B, C, D], E>
): Serializer<[A, B, C, D, E]>
export function concat(...args: ConcatArg<any, any>[]): Serializer<any> {
  return _concat(...args);
}