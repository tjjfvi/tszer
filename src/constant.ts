import { Serializer } from "./Serializer";

export const constant = <T>(value: T) => new Serializer<T>({
  length: 0,
  serialize: () => { },
  deserialize: () => value,
});