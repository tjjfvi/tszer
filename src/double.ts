import { Serializer } from "./Serializer";

export const doubleLE = () => new Serializer<number>({
  length: 8,
  serialize: (num, buf, off) => buf.writeDoubleLE(num, off),
  deserialize: (buf) => buf.readDoubleLE(0),
});

export const doubleBE = () => new Serializer<number>({
  length: 8,
  serialize: (num, buf, off) => buf.writeDoubleBE(num, off),
  deserialize: (buf) => buf.readDoubleBE(0),
});
