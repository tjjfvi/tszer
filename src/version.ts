import { Serializer } from "./Serializer"

export type Version = string | number | Buffer;

export const version = <V extends Version>(versions: V[]) => {
  const buffers = versions.map(versionToBuffer);
  checkVersionUnabiguity(buffers);
  return new Serializer<V>({
    serialize: (version, writeChunk) => writeChunk(versionToBuffer(version)),
    deserialize: async getChunk => {
      let pool = buffers.slice();
      let consumed = 0;
      let length = 0;
      while (pool.length && (length = Math.min(...pool.map(b => b.length)) - consumed)) {
        let chunk = await getChunk(length);
        pool = pool.filter(b => chunk.equals(b.slice(consumed, consumed + length)));
        consumed += length;
      }
      if (!pool.length)
        throw new Error("Invalid version");
      return versions[buffers.indexOf(pool[0])];
    }
  })
}

export const checkVersionUnabiguity = (buffers: Buffer[]) => {
  type Lookup = "end" | { [k: number]: Lookup } | undefined;
  let lookup: { [k: number]: Lookup } = { 0: undefined };
  for (let buffer of buffers) {
    let prevLookup = lookup;
    let key = 0;
    let ind = 0;
    while (true) {
      let curLookup = prevLookup[key];

      if (buffer.length === ind) {
        if (curLookup)
          throw new Error(ambiguousVersionErrorMessage(buffer, buffers));
        prevLookup[key] = "end";
        break;
      }

      curLookup = prevLookup[key] = curLookup ?? {};
      if (curLookup === "end")
        throw new Error(ambiguousVersionErrorMessage(buffer, buffers));
      prevLookup = curLookup;
      key = buffer[ind++];
    }
  }
}

export const versionToBuffer = (v: Version) =>
  v instanceof Buffer ?
    v.slice() :
    typeof v === "string" ?
      Buffer.from(v, "utf8") :
      Buffer.from([v])

const ambiguousVersionErrorMessage = (buffer: Buffer, buffers: Buffer[]) =>
  `Ambiguous version ${buffer.toString("hex")}\n` + buffers.map(b =>
    `  - ${b.toString("hex")}`
  ).join("\n")