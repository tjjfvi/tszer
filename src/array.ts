import { Serializer } from "./Serializer";
import { concat } from "./concat";
import { uint16LE } from "./uint";

export const repeated = <T>(t: Serializer<T>, length: number): Serializer<T[]> => {
  if (length === 1)
    return t.map({
      serialize: (arr) => {
        if (arr.length !== 1)
          throw new Error("repeated was passed invalid array of length !== 1");
        return arr[0];
      },
      deserialize: t => [t],
    });
  return concat(t, repeated(t, length - 1)).map({
    serialize: (arr) => {
      if (arr.length !== length)
        throw new Error(`repeated was passed invalid array of length !== ${length}`);
      return [arr[0], arr.slice(1)];
    },
    deserialize: ([val, arr]) => [val, ...arr],
  });
}

export const array = <T>(t: Serializer<T>, numberSerializer = uint16LE()) =>
  concat(
    numberSerializer,
    ([num]) => repeated(t, num),
  ).map<T[]>({
    serialize: val => [val.length, val],
    deserialize: ([, val]) => val,
  })
