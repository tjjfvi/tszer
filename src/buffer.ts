import { uint16LE } from "./uint";
import { concat } from "./concat";
import { Serializer } from "./Serializer";

export const constLengthBuffer = (length: number) => new Serializer<Buffer>({
  length,
  serialize: (val, buf, off) => buf.fill(val, off, off + length),
  deserialize: (buf, off) => buf.slice(off, off + length),
});

export const buffer = (numberSerializer = uint16LE()) => concat(
  numberSerializer,
  ([num]) => constLengthBuffer(num),
).map<Buffer>({
  serialize: buf => [buf.length, buf],
  deserialize: ([, buf]) => buf,
});