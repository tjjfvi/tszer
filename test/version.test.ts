import { Serializer } from "../src/Serializer";
import { version, versionToBuffer, checkVersionUnabiguity, Version } from "../src/version";
import { dss } from "./dss";
import { Readable } from "stream";

const validVersionTests: [string, string[]][] = [
  ["justEmptyString", [""]],
  ["singleLetters", ["a", "b", "c"]],
  ["commonPrefix", ["aa", "ab", "ac"]],
  ["long", ["this is a sentence", "and another sentence", "and finally this"]],
  ["complex", [
    "apx",
    "apy",
    "apz",
    "aqx",
    "aqy0",
    "aqyw",
    "b",
    "special characters: !@#$%^&*()-_=+[]\\/><,.~",
    "weird: \0\b\v\t\f\n\r\x0B\u1234\u5678",
  ]]
]

describe("version", () => {
  describe.each(validVersionTests)("%s", (name, versions) => {
    const serializer = version(versions);
    test.each(versions)("%s", async version => {
      expect(await dss(serializer, version)).toEqual(version)
    })
    describe("invalid", () => {
      test("long", async () => {
        await expect(
          serializer.deserialize(Readable.from(["this is always invalid"]))
        ).rejects.toThrowErrorMatchingSnapshot();
      })
      if (name !== "justEmptyString")
        test("emptyString", async () => {
          await expect(
            serializer.deserialize(Readable.from([]))
          ).rejects.toThrowErrorMatchingSnapshot();
        })
    })
  })
});

describe("versionToBuffer", () => {
  test.each<[string, Version]>([
    ["emptyString", ""],
    ["string", "abc123"],
    ["buffer", Buffer.from([0x01, 0x18, 0x99, 0x98, 0x81, 0x99, 0x91, 0x19, 0x72, 0x53])],
    ["0", 0],
    ["1", 1],
    ["2", 2],
    ["255", 255],
  ])("%s", (_, version) => {
    expect(versionToBuffer(version)).toMatchSnapshot();
  })
})

describe("checkVersionUnambiguity", () => {
  describe("valid", () => {
    test.each(validVersionTests)("%s", (_, versions) => {
      const buffers = versions.map(versionToBuffer);
      expect(() => checkVersionUnabiguity(buffers)).not.toThrow();
    })
  })

  describe("invalid", () => {
    test.each<[string, string[]]>([
      ["emptyStringWithMultiple", ["", "a"]],
      ["singleLetters", ["a", "b", "c", "c"]],
      ["commonPrefix", ["aa", "ab", "ac", "acid"]],
      ["long", ["this is a sentence", "and another sentence", "and another sentence", "and finally this"]],
      ["complex", [
        "ap",
        "apx",
        "apy",
        "apya",
        "apz",
        "aqx",
        "aqy0",
        "apy0-prime",
        "aqyw",
        "b",
        "special characters: !@#$%^&*()-_=+[]\\/><,.~",
        "special characters: !@#$%^&*()-_=+[]\\/><,.~~~~~~~~~~~~~~~~~~~~~~~~~~~~",
        "weird: \0\b\v\t\f\n\r\x0B\u1234\u5678",
        "weird: \0\b\v\t\f\n\r\x0B\u1234\u5678\x42",
      ]],
    ])("%s", (_, versions) => {
      const buffers = versions.map(versionToBuffer);
      expect(() => checkVersionUnabiguity(buffers)).toThrowErrorMatchingSnapshot();
    })
  })
})
