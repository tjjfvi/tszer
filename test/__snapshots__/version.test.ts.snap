// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkVersionUnambiguity invalid commonPrefix 1`] = `
"Ambiguous version 61636964
  - 6161
  - 6162
  - 6163
  - 61636964"
`;

exports[`checkVersionUnambiguity invalid complex 1`] = `
"Ambiguous version 617078
  - 6170
  - 617078
  - 617079
  - 61707961
  - 61707a
  - 617178
  - 61717930
  - 617079302d7072696d65
  - 61717977
  - 62
  - 7370656369616c20636861726163746572733a2021402324255e262a28292d5f3d2b5b5d5c2f3e3c2c2e7e
  - 7370656369616c20636861726163746572733a2021402324255e262a28292d5f3d2b5b5d5c2f3e3c2c2e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e7e
  - 77656972643a2000080b090c0a0d0be188b4e599b8
  - 77656972643a2000080b090c0a0d0be188b4e599b842"
`;

exports[`checkVersionUnambiguity invalid emptyStringWithMultiple 1`] = `
"Ambiguous version 61
  - 
  - 61"
`;

exports[`checkVersionUnambiguity invalid long 1`] = `
"Ambiguous version 616e6420616e6f746865722073656e74656e6365
  - 7468697320697320612073656e74656e6365
  - 616e6420616e6f746865722073656e74656e6365
  - 616e6420616e6f746865722073656e74656e6365
  - 616e642066696e616c6c792074686973"
`;

exports[`checkVersionUnambiguity invalid singleLetters 1`] = `
"Ambiguous version 63
  - 61
  - 62
  - 63
  - 63"
`;

exports[`version commonPrefix invalid emptyString 1`] = `"Unexpected end of input"`;

exports[`version commonPrefix invalid long 1`] = `"Invalid version"`;

exports[`version complex invalid emptyString 1`] = `"Unexpected end of input"`;

exports[`version complex invalid long 1`] = `"Invalid version"`;

exports[`version justEmptyString invalid long 1`] = `"Expected end of input"`;

exports[`version long invalid emptyString 1`] = `"Unexpected end of input"`;

exports[`version long invalid long 1`] = `"Invalid version"`;

exports[`version singleLetters invalid emptyString 1`] = `"Unexpected end of input"`;

exports[`version singleLetters invalid long 1`] = `"Invalid version"`;

exports[`versionToBuffer 0 1`] = `
Object {
  "data": Array [
    0,
  ],
  "type": "Buffer",
}
`;

exports[`versionToBuffer 1 1`] = `
Object {
  "data": Array [
    1,
  ],
  "type": "Buffer",
}
`;

exports[`versionToBuffer 2 1`] = `
Object {
  "data": Array [
    2,
  ],
  "type": "Buffer",
}
`;

exports[`versionToBuffer 255 1`] = `
Object {
  "data": Array [
    255,
  ],
  "type": "Buffer",
}
`;

exports[`versionToBuffer buffer 1`] = `
Object {
  "data": Array [
    1,
    24,
    153,
    152,
    129,
    153,
    145,
    25,
    114,
    83,
  ],
  "type": "Buffer",
}
`;

exports[`versionToBuffer emptyString 1`] = `
Object {
  "data": Array [],
  "type": "Buffer",
}
`;

exports[`versionToBuffer string 1`] = `
Object {
  "data": Array [
    97,
    98,
    99,
    49,
    50,
    51,
  ],
  "type": "Buffer",
}
`;
